### PR TITLE
ActionBasedMessagingExtension Scenario

### PR DESCRIPTION
**Pull request by stevengum**
For comparison to #2579

### CODE REVIEW

File: tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs [hunk 1]
```diff
@@ -17,5 +18,104 @@ protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivi
         {
             await turnContext.SendActivityAsync(MessageFactory.Text($"echo: {turnContext.Activity.Text}"), cancellationToken);
         }
+
+        protected override async Task<MessagingExtensionActionResponse> OnTeamsMessagingExtensionSubmitActionAsync(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action, CancellationToken cancellationToken)
+        {
+            switch (action.CommandId)
+            {
+                case "createCard":
```
Comment (johnataylor) [inline tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs]: Could we have a comment here saying these strings come from the manifest.json

File: tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs [hunk 2]
```diff
@@ -17,5 +18,104 @@ protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivi
...
+        }
+
+        private MessagingExtensionActionResponse HandleShareMessageCommand(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action)
+        {
+            var heroCard = new HeroCard
+            {
+                Title = $"{action.MessagePayload.From.User.DisplayName} sent this message:",
+                Text = action.MessagePayload.Body.Content,
+            };
+            var includeImageData = (action.Data as JObject).ToObject<IncludeImageData>();
```
Comment (johnataylor) [inline tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs]: Using a () style cast operator will result in an exception that the object can't be cast to this type. Using the "as" operator style cast gives you a more cryptic null exception.

File: tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs [hunk 3]
```diff
@@ -17,5 +18,104 @@ protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivi
...
+            }
+
+            var response = new MessagingExtensionActionResponse();
+            response.ComposeExtension = new MessagingExtensionResult
+            {
+                Type = "result",
+                AttachmentLayout = "list",
+                Attachments = new List<MessagingExtensionAttachment>
+                {
+                    new MessagingExtensionAttachment
```
Comment (johnataylor) [inline tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs]: This is slightly ridiculous. At the very least a ToMessagingExtensionAttachment(). But ideally we could do better than that.

Comment (stevengum) [reply tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs]: Resolving as issue #2603 was filed to address this feedback.

File: tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs (hunk 2 above)
Comment (stevengum) [reply tests/Teams/ActionBasedMessagingExtension/Bots/ActionBasedMessagingExtensionBot.cs]: Done

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

**Pull request merged**